### PR TITLE
Account email search API

### LINKED ISSUE #97: [Access] [BE] Account email search API
## Context

On the access page, admins should be able to search accounts by email so that they can easily revoke or update a user's access to the tool instead of having to sift through all emails on the page.

## Requirements

Create a new API route using the `account` router. This route should live at `account.search` and take in an input of:

- searchSubject: string

The API route should perform an infix search on the email field in the `accounts` MongoDB collection with the search term (e.g. if the search term was abc, then `%abc%`). Use regex for this. You will need to create another database action to support this search. If there is a database error, throw a useful tRPC error from the API.

#### PM

- [ ] Product-specific requirements
- [ ] I should be able to search for an email using this API and it does not need to start with what I use to sear (as I could search "jane" and I want the email "123janeiscool@gmail.com" to pop up

#### EM

- [ ] Output of the API route shall be all documents that match the regex pattern
- [ ] Write unit tests for the db action and API route you created
- [ ] Errors handled with a useful message

### PR DESCRIPTION
**Pull request by james00100**
## PR Title

Issue Number(s): 97

What does this PR change and why?

### Checklist

- [ ] Requirements have been implemented
- [ ] Acceptance criteria are met
- [ ] Database schema docs have been updated or are not necessary
- [ ] Code follows design and style guidelines
- [ ] Commits follow guidelines (concise, squashed, etc)
- [x] Relevant reviewers (Senior Dev/EM/Designers) have been assigned to this PR

### Critical Changes

- Database change/migration to run
- Environment config change
- Breaking API change

### Related PRs

- #number_of_pr

### Testing

Enumerate steps to test the functionality of your ticket. This should include edge cases and testing of error handling, if applicable.

### CODE REVIEW

**Review by katsukixyz — COMMENTED** [review]
Database action looks good, make sure you have a corresponding unit test in`__tests__/db/actions/Account.unit.test.ts`. This ticket also requires that you make an API endpoint at `account.search` that will use this database action - make sure you have a unit tests for that too in `__tests__/pages/api/account/account.unit.ts`.

File: web/server/routers/account.ts [hunk 1]
```diff
@@ -7,16 +7,22 @@ import {
   findAccount,
   findAll,
   removeAllAccounts,
+  searchAccount,
 } from "../../db/actions/Account";
 import { updateAllUsers, updateUserByEmail } from "../../db/actions/User";
 import Account from "../../db/models/Account";
 import { IAccount, Role } from "../../utils/types/account";
 import { router, procedure } from "../trpc";
+import { Search2Icon } from "@chakra-ui/icons";
```
Comment (katsukixyz) [inline web/server/routers/account.ts]: Unused import

File: web/db/actions/Account.ts [hunk 2]
```diff
@@ -99,4 +99,27 @@ async function findAll(
...
+    const regexTerm = new RegExp(`.*${searchSubject}.*`, 'i');
+    return await Account.findOne(
+      { email: regexTerm },
+      { _id: 0, __v: 0 },
+      { session: session }
+    );
+  } catch (e) {
+    return null;
+  }
+}
```
Comment (katsukixyz) [inline web/db/actions/Account.ts]: This action needs to return _all_ documents that match the regex pattern. Update tests accordingly. Output should be an array unless there is a database error, in which case return null

File: web/server/routers/account.ts [hunk 3]
```diff
@@ -207,4 +213,13 @@ export const accountRouter = router({
...
     }
   }),
+
+  search: procedure
+    .input(searchInput)
+    .query(async ({ input, ctx }) => {
+      const session = await Account.startSession();
+      const { searchSubject } = input;
+      const account = await searchAccount(searchSubject, session);
+      return account;
```
Comment (katsukixyz) [inline web/server/routers/account.ts]: Include a try catch that throws a TRPCError if there's an issue with the transactions. Also handle the case where `searchAccounts` returns null, which means there was a db issue

File: web/db/actions/Account.ts [hunk 4]
```diff
@@ -99,4 +99,27 @@ async function findAll(
...
 
-export { findAccount, addAccount, removeAllAccounts, updateAccount, findAll };
+/**
+ * Finds account by search term using regex
+ *
+ * @param searchSubject string in regex
+ * @param session MongoDB session used for transactions
+ * @returns pruned document (without _id, __v) if successful, null otherwise
+ */
+async function searchAccount(
```
Comment (katsukixyz) [inline web/db/actions/Account.ts]: Rename to `searchAccounts`

**Review by katsukixyz — CHANGES_REQUESTED** [review]
Sorry for not catching the database action output shape in my previous review. You'll also need to write a test for the database action itself, `__tests__/db/actions/Account.ts`. Looking good!

File: web/server/routers/account.ts [hunk 5]
```diff
@@ -218,8 +217,25 @@ export const accountRouter = router({
     .input(searchInput)
     .query(async ({ input, ctx }) => {
       const session = await Account.startSession();
```
Comment (katsukixyz) [inline web/server/routers/account.ts:219]: Nit: probably unnecessary to use a transaction here since we are only making 1 database mutation.

**Review by katsukixyz — COMMENTED** [review]
Nice work! It seems like the test runner is failing - could you take a look? We have a `yarn jest` command that you can use to run our tests. Everything else looks awesome. Will approve once the tests are working in our pipeline

File: web/__tests__/db/actions/Account.unit.test.ts [hunk 6]
```diff
@@ -185,4 +186,26 @@ describe("[DB] Account - Unit Test", () => {
...
   });
+
+  describe("searchAccounts", () => {
+    test("happy", async () => {
+      const searchTerm = randomAccounts[0].email;
+      const accounts = await searchAccounts(searchTerm);
+  
+      expect(accounts).not.toBeNull();
+      expect(accounts).toBeInstanceOf(Array);
+      if (accounts === null) return;
```
Comment (katsukixyz) [inline web/__tests__/db/actions/Account.unit.test.ts]: Sorry for not catching this earlier - you can remove this line

File: web/__tests__/db/actions/Account.unit.test.ts [hunk 7]
```diff
@@ -185,4 +186,26 @@ describe("[DB] Account - Unit Test", () => {
...
+      expect(accounts.every((account) => account.email.includes(searchTerm))).toBe(true);
+    });
+  
+    test("not found", async () => {
+      const searchTerm = "ASDASDASD";
+      const accounts = await searchAccounts(searchTerm);
+  
+      expect(accounts).not.toBeNull();
+      expect(accounts).toBeInstanceOf(Array);
+      if (accounts === null) return;
```
Comment (katsukixyz) [inline web/__tests__/db/actions/Account.unit.test.ts]: See above

**Review by katsukixyz — COMMENTED** [review]
Minor comments that I should've caught in earlier reviews, sorry about that. Sending this to the rest of the LT in the meantime for review

**Review by katsukixyz — APPROVED** [review]
Great work! I'll merge it in.
